crates/neon/src/types_impl/function/mod.rs

```diff
@@ -7,11 +7,75 @@ use crate::{
     handle::Handle,
     object::Object,
     result::{JsResult, NeonResult},
-    types::{JsFunction, JsObject, JsValue, Value},
+    types::{
+        extract::{TryFromJs, TryIntoJs},
+        JsFunction, JsObject, JsValue, Value,
+    },
 };
 
 pub(crate) mod private;
 
+/// A builder for making a JavaScript function call like `parseInt("42")`.
+///
+/// The builder methods make it convenient to assemble the call from parts:
+/// ```
+/// # use neon::prelude::*;
+/// # fn foo(mut cx: FunctionContext) -> JsResult<JsNumber> {
+/// # let parse_int: Handle<JsFunction> = cx.global("parseInt")?;
+/// let x: f64 = parse_int
+///     .bind(&mut cx)
+///     .arg("42")?
+///     .apply()?;
+/// # Ok(cx.number(x))
+/// # }
+/// ```
+pub struct BindOptions<'a, 'cx: 'a, C: Context<'cx>> {
+    pub(crate) cx: &'a mut C,
+    pub(crate) callee: Handle<'cx, JsFunction>,
+    pub(crate) this: Option<Handle<'cx, JsValue>>,
+    pub(crate) args: private::ArgsVec<'cx>,
+}
+
+impl<'a, 'cx: 'a, C: Context<'cx>> BindOptions<'a, 'cx, C> {
+    /// Set the value of `this` for the function call.
+    pub fn this<V: Value>(&mut self, this: Handle<'cx, V>) -> &mut Self {
+        self.this = Some(this.upcast());
+        self
+    }
+
+    /// Replaces the arguments list with the given arguments.
+    pub fn args<A: TryIntoArguments<'cx>>(&mut self, a: A) -> NeonResult<&mut Self> {
+        self.args = a.try_into_args_vec(self.cx)?;
+        Ok(self)
+    }
+
+    /// Add an argument to the arguments list.
+    pub fn arg<A: TryIntoJs<'cx>>(&mut self, a: A) -> NeonResult<&mut Self> {
+        let v = a.try_into_js(self.cx)?;
+        self.args.push(v.upcast());
+        Ok(self)
+    }
+
+    /// Add an argument to the arguments list, computed from a closure.
+    pub fn arg_with<R, F>(&mut self, f: F) -> NeonResult<&mut Self>
```

Comment:
Should there be a version of this function that returns `R: TryIntoArguments`?

I think *no* because a user should _probably_ just create this tuple before calling `.bind(..)`. It's more API noise without a valuable use case.

Reply:
Do you mean something like:
```rust
pub fn args_with<R, F>(&mut self, f: F) -> NeonResult<&mut Self>
where
    R: TryIntoArguments<'cx>,
    F: FnOnce(&mut Cx<'cx>) -> R,
{
    // ...
}
```
?

Reply:
Resolved: needs another impl for `With` on `TryIntoArguments`, and the actual implementation of `.args_with` could be to delegate to `.args(With(...))`.